src/Workspaces/Core/Portable/Utilities/ReferenceHolder`1.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Roslyn.Utilities
+{
+    internal readonly struct ReferenceHolder<T> : IEquatable<ReferenceHolder<T>>
+        where T : class?
+    {
+        [AllowNull, MaybeNull]
+        private readonly T _strongReference;
+        private readonly WeakReference<T>? _weakReference;
+        private readonly int _hashCode;
+
+        private ReferenceHolder(T strongReference)
+        {
+            _strongReference = strongReference;
+            _weakReference = null;
+            _hashCode = 0;
+        }
+
+        private ReferenceHolder(WeakReference<T> weakReference, int hashCode)
+        {
+            _strongReference = null;
+            _weakReference = weakReference;
+            _hashCode = hashCode;
+        }
+
+        public static ReferenceHolder<T> Strong(T value)
+            => new ReferenceHolder<T>(value);
+
+        public static ReferenceHolder<T> Weak(T value)
+        {
+            if (value is null)
+            {
+                // Track this as a strong reference so we know 'Equals' should only look at values originally null.
+                return Strong(value);
+            }
+
+            return new ReferenceHolder<T>(new WeakReference<T>(value), RuntimeHelpers.GetHashCode(value));
+        }
+
+        [return: MaybeNull]
+        public T TryGetTarget()
+        {
+            if (_weakReference is object)
+                return _weakReference.GetTarget();
+
+            return _strongReference;
+        }
+
+        public override bool Equals(object? obj)
+        {
+            return obj is ReferenceHolder<T> other
+                && Equals(other);
+        }
+
+        public bool Equals(ReferenceHolder<T> other)
+        {
+            var x = TryGetTarget();
+            var y = other.TryGetTarget();
+            if (x is null)
+            {
+                if (_weakReference is object)
+                {
+                    // 'x' is a weak reference that was collected. Verify 'y' is a collected weak reference with the
+                    // same runtime hash code.
+                    return y is null && other._weakReference is object && _hashCode == other._hashCode;
```

Comment:
that seems highly suspect to me... 

this def makes me want for the "WeakSet" that simply returns true/false about if the value is the set at the point of asking.

This seems to be having the semantics of "the value may have been there *at some point in the past* with some degree of confidence".  i.e. it's as accurate as hash-collisions are accurate.  Note: i believe the compiler has cases where they hash-collide on non-equal stuff just for simplicity of the hash calculation.

Reply:
> ... as hash-collisions are accurate ...

I switched to `RuntimeHelpers.GetHashCode`. It is impossible for collisions to occur when using `WeakSet<T>` because one of the references is alive at the time of the check.

Reply:
RuntimeHelpers.GetHashCode always returning 42 is a perfectly valid return as far as I'm concerned. It's a poorly performing implementation no doubt, but probably legal. As far as I'm aware GetHashCode is effectively the address of the sync block, but that's absolutely an implementation detail.

That said I don't think this matters and I think the code works but it's absolutely subtle.

Reply:
➡️ Updated the documentation here

Reply:
> ... effectively the address of the sync block ...

This wouldn't work for compacting GC (runtime hash code is not allowed to change after it is first observed).

Reply:
i guess i don't understand.  if *either* side is collected, i woudl expect the WeakSet to return 'false' for equals.  They're just not equal.

Reply:
That would allow cases where the comparison of two equal and immutable objects returns true, followed by the same call returning false.

Reply:
*yes*.  That si my expected semantics when it comes to a weakset (and weakrefs).  i absolutely think it can (and should) be able to return 'true' on one call, and then 'false' afterwards.

Reply:
The approach is problematic for use in sets. Even though we don't perform `Remove` operations in this case, it's a pretty big red flag for any future reviewer that sees a weak set implementation. The approach here is the lesser evil in terms of likelihood of accidental incorrect behavior.

Reply:
i honestly strongly disagree.  I can't see how this is correct all.  It's completely feasible for thsi to produce entirely incorrect results.  

It's a *weak* set.  If either side got GC'ed i would *aboslutely* expect the set to say "no, it's not in ehre".  Furthermore, as this type exists only for WeakSet, i don't think it should contain any semantics we don't need there.  And we should likely make this type a nested type of WeakSet.